src/os/bluestore/BlueStore.cc

```diff
@@ -6555,8 +6636,154 @@ int BlueStore::_read_bdev_label(CephContext* cct, const string &path,
   return 0;
 }
 
+/**
+  Reads device label.
+  *out_label      - Filled if reading of label is considered successful.
+  *out_valid_positions - List of locations that contained valid labels.
+  *out_is_multi  -  Whether the label is regular or multi label with epoch.
+  *out_epoch     -  Epoch of label.
+
+  Returns:
+  0    When all label are read
+  1    When some, but not all labels are read
+  -ENOENT Otherwise
+
+*/
+int BlueStore::_read_main_bdev_label(
+  bluestore_bdev_label_t *out_label,
+  std::vector<uint64_t>* out_valid_positions,
+  bool* out_is_multi,
+  int64_t* out_epoch)
+{
+  dout(10) << __func__ << dendl;
+  ceph_assert(out_label);
+  // go and try read all possible bdev labels.
+  // if only first bdev label is correct, it must not have "multi=yes" key.
+  int64_t epoch = -1;
+  bool all_labels_valid = true;
+  for (uint64_t position : bdev_label_positions) {
+    bluestore_bdev_label_t label;
+    int r = _read_bdev_label(cct, path + "/block", &label, position);
+    if (r == 0 && (fsid.is_zero() || label.osd_uuid == fsid)) {
+      auto i = label.meta.find("multi");
+      bool is_multi = i != label.meta.end() && i->second == "yes";
+      if (position == BDEV_FIRST_LABEL_POSITION && !is_multi) {
+        // we have a single-label case
+        *out_label = label;
+        is_multi = false;
+        if (out_is_multi) {
+          *out_is_multi = false;
+        }
+        if(out_valid_positions) {
+          out_valid_positions->push_back(position);
+        }
+        goto done;
+      }
+      if (!is_multi) {
+        // for not base bdev position, it has to be cloned to be considered
+        continue;
+      }
+      i = label.meta.find("epoch");
+      if (i != label.meta.end()) {
+        int64_t v = atoll(i->second.c_str());
+        if (v > epoch) {
+          epoch = v;
+          *out_label = label;
+          if (out_epoch) {
+            *out_epoch = epoch;
+          }
+          is_multi = true;
+          if (out_is_multi) {
```

Comment:
No need to do that in a loop, can be done out of loop before 'done' label.

Reply:
moved.